edit_course/operations/configure.py

```diff
@@ -356,299 +357,357 @@ def get_build_log(instance):
-    category_map = {}
-    seen = []
-    for key, c in config.get("categories", {}).items():
-        if not "name" in c:
-            errors.append(_('COURSE_CONFIG_ERROR_CATEGORY_REQUIRES_NAME'))
-            continue
-        try:
-            category = instance.categories.get(name=format_localization(c["name"]))
-        except LearningObjectCategory.DoesNotExist:
-            category = LearningObjectCategory(course_instance=instance,
-                name=format_localization(c["name"]))
-        if "status" in c:
-            category.status = str(c["status"])
-        if "description" in c:
-            category.description = str(c["description"])
-        if "points_to_pass" in c:
-            i = parse_int(c["points_to_pass"], errors)
-            if not i is None:
-                category.points_to_pass = i
-        for field in [
-            "confirm_the_level",
-            "accept_unofficial_submits",
-        ]:
-            if field in c:
-                setattr(category, field, parse_bool(c[field]))
-        category.full_clean()
-        category.save()
-        category_map[key] = category
-        seen.append(category.id)
-
-    for category in instance.categories.all():
-        if not category.id in seen:
-            category.status = LearningObjectCategory.STATUS.HIDDEN
-            category.save()
-
-    # Configure course modules.
-    seen_modules = []
-    seen_objects = []
-    nn = 0
-    n = 0
-    for m in config.get("modules", []):
-        if not "key" in m:
-            errors.append(_('COURSE_CONFIG_ERROR_MODULE_REQUIRES_KEY'))
-            continue
-        try:
-            module = instance.course_modules.get(url=str(m["key"]))
-        except CourseModule.DoesNotExist:
-            module = CourseModule(course_instance=instance, url=str(m["key"]))
-
-        if "order" in m:
-            module.order = parse_int(m["order"], errors)
-        else:
-            n += 1
-            module.order = n
-
-        if "title" in m:
-            module.name = format_localization(m["title"])
-        elif "name" in m:
-            module.name = format_localization(m["name"])
-        if not module.name:
-            module.name = "-"
-        if "status" in m:
-            module.status = str(m["status"])
-        if "points_to_pass" in m:
-            i = parse_int(m["points_to_pass"], errors)
-            if not i is None:
-                module.points_to_pass = i
-        if "introduction" in m:
-            module.introduction = str(m["introduction"])
-
-        if "open" in m:
-            dt = parse_date(m["open"], errors)
-            if dt:
-                module.opening_time = dt
-        if not module.opening_time:
-            module.opening_time = instance.starting_time
+        return False, errors
 
-        if "close" in m:
-            dt = parse_date(m["close"], errors)
+    # wrap everything in a transaction to make sure invalid configuration isn't saved
+    with transaction.atomic():
```

Comment:
The diff includes everything since all lines were indented. Did you change anything besides the "publish_url" in the end?

Reply:
I added rollbacks to the error points "transaction.set_rollback(True)"

Reply:
That's good!